9.5/paper.tex

```diff
@@ -490,6 +491,41 @@ \subsubsection{Interfaces to Trilinos' Belos and NOX packages}\label{sec:trilino
   elaborate? At least in the KINSOL interfaces, you only update the
   preconditioner whenever setup\_jacobian() is called.}
 
+\subsubsection{Uniform interface for nonlinear solvers}\label{sec:nonlinear}
+
+With the current release, \dealii{} now supports solvers for nonlinear problems
+provided by several different external packages:
+1. KINSOL (part of SUNDIALS), 2. SNES (part of PETSc, see Section~\ref{sec:petsc}), 3. NOX (part of Trilinos, see Section~\ref{sec:trilinos}).
+The wrappers are provided by the classes \texttt{PETScWrappers::NonlinearSolver<VectorType>}, \texttt{SUNDIALS::KINSOL<VectorType>}, and \texttt{TrilinosWrappers::NOXSolver<VectorType>}, respectively.
+
+All classes have a very similar interface, except that they vary in the kind of
+algorithms and parameters offered. The new class \texttt{NonlinearSolverSelector<VectorType>} provides a wrapper on top of
+the three external solvers with a unified interface. The user can either let
+\dealii{} decide which of the packages to use (depending on the current availability of
+the external packages) or specify it manually.
+
+Here is a complete example of an MPI parallel Newton solver that automatically chooses one of the available packages:
+\begin{c++}
+    using NLSolver = NonlinearSolverSelector<LA::MPI::Vector>;
+
+    NLSolver::AdditionalData additional_data(NLSolver::automatic,
+                                             NLSolver::newton);
+    NLSolver nonlinear_solver(additional_data, mpi_communicator);
+
+    nonlinear_solver.reinit_vector = [&](LA::MPI::Vector &x)
+                                     {/*...*/};
+    nonlinear_solver.residual = [&](const LA::MPI::Vector &evaluation_point,
+                                    LA::MPI::Vector &      residual)
+                                {/*...*/};
+    nonlinear_solver.setup_jacobian = [&](const LA::MPI::Vector &current_u)
+                                      {/*...*/};
+    nonlinear_solver.solve_with_jacobian = [&](const LA::MPI::Vector &rhs,
+                                               LA::MPI::Vector &      dst,
+                                               const double tolerance)
+                                           {/*...*/};
+
+    nonlinear_solver.solve(current_solution);
+\end{c++}
```

Comment:
Will this actually ever use PETSc? It's not a PETSc vector type, and I thought that SNES requires a PETSc vector type.

Reply:
It can use SNES provided the vector type is a PETSc Vector, yes.